lifecycle.Rmd

```diff
@@ -65,78 +264,193 @@ usethis::use_version()
+You can get much more detail in `vignette("stages", package = "lifecycle")`.
 
-    ```{r}
-    bar <- function(x, y, z) {
-      if (!missing(y)) {
-        warning("argument y is deprecated; please use z instead.", 
-          call. = FALSE)
-        z <- y
-      }
-    }
+The lifecycle stage is most often communicated through a badge.
+If you'd like to use lifecycle badges, call `usethis::use_lifecycle()` to do some one-time setup:
 
-    bar(1, 2, 3)
-    ```
+```{r}
+#| include: false
+#| cache: false
+temp_pkg <- fs::path_temp("somepackage")
+withr::with_options(
+  list(usethis.description = NULL),
+  usethis::create_package(
+    temp_pkg, open = FALSE, rstudio = TRUE, check_name = FALSE
+  )
+)
+op <- usethis::proj_get()
+usethis::proj_set(temp_pkg)
+usethis::use_package_doc(open = FALSE)
+```
+
+```{r}
+usethis::use_lifecycle()
+```
+
+```{r}
+#| include: false
+usethis::proj_set(op)
+unlink(temp_pkg)
+```
+
+This leaves you in a position to use lifecycle badges in help topics (and to use lifecycle functions, which we discuss in ??).
+For a function, include the badge in its `@description` block.
+Here's how we indicate that `dplyr::top_n()` is "superseded":
+
+```{r}
+#' Select top (or bottom) n rows (by value)
+#'
+#' @description
+#' `r lifecycle::badge("superseded")`
+#' `top_n()` has been superseded in favour of [slice_min()]/[slice_max()]....
+```
+
+For a function argument, include the badge in the `@param` tag.
+Here's how the deprecation of `readr::write_file(path =)` is indicated:
+
+```{r}
+#' @param path `r lifecycle::badge("deprecated")` Use the `file` argument
+#'   instead.
+```
+
+Call `usethis::use_lifecycle_badge()` if you want to use a badge in README to indicate the lifecycle of an entire package.
+
+If the lifecycle of a package is "stable", we might not bother with a badge, as "stable" is the assumed default.
+Similarly, we typically only use a badge for a function or argument if its stage differs from the overall stage of the package.
+
+### Deprecating a function or argument
+
+If you're going to remove a function or argument, it's usually best to do so in phases.
+Deprecation is a general term for the situation where something is explicitly discouraged, but it has not yet been removed.
+
+The `lifecycle::deprecate_warn()` function can be used inside a function to inform your user that they're using a deprecated feature and, ideally, to let them know about the preferred alternative.
+
+```{r R.options = list("lifecycle:::calling_package" = "somepackage", "lifecycle_verbosity" = "warning")}
+# new function
+add3 <- function(x, y, z) {
+  x + y + z
+}
+
+# old function
+plus3 <- function(x, y, z) {
+  lifecycle::deprecate_warn("1.0.0", "plus3()", "add3()")
+  add3(x, y, z)
+}
+
+plus3(1, 2, 3)
+```
+
+At this point, a user who calls `plus3()` sees a warning explaining that the function has a new name, but we go ahead and call `add3()` with their inputs.
```

Comment:
Because warning every time a function is used is very annoying (almost as annoying as removing the function), by default `deprecate_warn()` will only warn once every 8 hours. You can make it warn on every use by setting `always = TRUE`.

Lifecycle is trying to walk a tightrope — it's goal is to be just annoying enough to get the user to fix the code before the function or argument goes away, but not annoying enough that it substantially hampers their quality of life.

Reply:
Incorporated